django/db/backends/base/introspection.py

```diff
@@ -36,6 +36,26 @@ def identifier_converter(self, name):
         """
         return name
 
+    def only_nonempty_tables(self, tables):
+        """
+        Take a list of tables, return a list of tables that have at least
+        one row.
+        """
+        if not tables:
+            return []
+        sql_list = []
+        for table in tables:
+            sql = """( SELECT '{}' AS has_rows FROM {} {} )""".format(
+                table, self.connection.ops.quote_name(table),
+                self.connection.ops.limit_offset_sql(0, 1)
+            )
+            sql_list.append(sql)
+        sql = " UNION ".join(sql_list)
+        with self.connection.cursor() as cursor:
+            cursor.execute(sql)
+            rows = cursor.fetchall()
+        return [row[0] for row in rows]
```

Comment:
We can do much better here. We don't need to regenerate the `LIMIT 1` snippet for every table, as it never changes. The `AS has_rows` is unnecessary.


```suggestion
        limit = self.connection.ops.limit_offset_sql(0, 1)
        sql = ' UNION '.join([
            "(SELECT '%s' FROM %s %s)" % (table, self.connection.ops.quote_name(table), limit)
            for table in tables
        ])
        with self.connection.cursor() as cursor:
            cursor.execute(sql)
            return [row[0] for row in cursor.fetchall()]
```

I'm also wondering whether we should pass the table names string literals in the `SELECT` as params to avoid any sniff of SQL injection. I honestly don't know if there is a real risk here. The advantage of keeping it like this is that we don't need to batch for SQLite and the overridden version there can be removed.

Reply:
You raise an interesting point here. Today I learned SQLite table names can be anything that doesn't start with `sqlite_` . For example:
```
sqlite> create table 'ab"''cd'(x integer);
sqlite> create table "ab'""cd"(x integer);
sqlite> .tables
ab"'cd  ab'"cd
```

Adding
```
class WeirdTable(models.Model):
    foo = models.TextField()

    class Meta:
        db_table = "ab'\"cd"
```
to `django/tests/introspection/models.py` breaks the tests. I guess people aren't crazy enough to put quotes in their table names?

Reply:
Haha. Well… 😅

I've done the following before to hook up the ORM to a view in a separate schema in PostgreSQL:

```python
    class Meta:
        managed = False
        db_table = 'schema"."matview'
```

Yay for dodgy hacks and workarounds! 🙈 

Slightly different to embedded single quotes though. Yuck. Hopefully nobody has tried to do that on purpose.

Reply:
Unforunately the table names string literals can't be params to `cursor.execute`. In my opinion, the issue of Django not quoting table names as demonstrated above is a bug and should be a different ticket. Thanks for your review, `skip_empty` is much better. Naming is hard!